plat/common/arm/mte.c

```diff
@@ -0,0 +1,117 @@
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <arm/arm64/cpu.h>
+#include <stddef.h>
+#include <uk/plat/mte.h>
+#include <uk/arch/lcpu.h>
+#include <uk/arch/random.h>
+#include <uk/arch/traps.h>
+#include <uk/event.h>
+
+void *mte_tag_contig(void *ptr, size_t size)
+{
+	uint64_t addr = (uint64_t)ptr;
+
+	UK_ASSERT(!(size % MTE_TAG_GRANULE));
+
+	addr = mte_insert_random_tag(addr);
+
+	for (unsigned int i = 0; i < size; i += MTE_TAG_GRANULE)
+		mte_store_alloc(addr + i);
+
+	return (void *)addr;
+}
+
+int mte_init(void)
+{
+	__u64 reg;
+	__u64 seed;
+	unsigned int mte_version;
+
+	mte_version = (SYSREG_READ(ID_AA64PFR1_EL1) >>
+			ID_AA64PFR1_EL1_MTE_SHIFT) &
+			ID_AA64PFR1_EL1_MTE_MASK;
+
+	if (mte_version < ARM64_FEAT_MTE2) {
+		uk_pr_err("FEAT_MTE2 is not implemented\n");
+		return -1;
+	}
+
+#if CONFIG_ARM64_FEAT_MTE_TCF_ASYMMETRIC
+	if (mte_version < ARM64_FEAT_MTE3) {
+		uk_pr_err("FEAT_MTE3 is not implemented\n");
+		return -1;
+	}
+#endif
+
+#if CONFIG_ARM64_FEAT_MTE_TCF_ASYNC
+	SYSREG_WRITE(SCTLR_EL1, (SYSREG_READ(SCTLR_EL1) |
+		       (MTE_TCF_ASYNC << SCTLR_EL1_TCF_SHIFT)));
+#elif CONFIG_ARM64_FEAT_MTE_TCF_ASYMMETRIC
+	SYSREG_WRITE(SCTLR_EL1, (SYSREG_READ(SCTLR_EL1) |
+		       (MTE_TCF_ASYMMETRIC << SCTLR_EL1_TCF_SHIFT)));
+#else
+	SYSREG_WRITE(SCTLR_EL1, (SYSREG_READ(SCTLR_EL1) |
+		       (MTE_TCF_SYNC << SCTLR_EL1_TCF_SHIFT)));
+#endif
+	/* Enable TBI */
+	SYSREG_WRITE(TCR_EL1, (SYSREG_READ(TCR_EL1) | TCR_EL1_TBI0_BIT));
+
+	/* TCR_EL1.TCMA0 must be zero */
+	SYSREG_WRITE(TCR_EL1, (SYSREG_READ(TCR_EL1) & ~TCR_EL1_TCMA0_BIT));
+
+	/* Use default random tag generation */
+	SYSREG_WRITE(GCR_EL1, (SYSREG_READ(GCR_EL1) & GCR_EL1_RRND_BIT));
+
+	/* Set seed for RGSR_EL1 */
+	if (ukarch_random_init() == -1)
```

Comment:
This is not safe according to the function description which might return any negative number to indicate an error.
```suggestion
	if (ukarch_random_init())
```

Reply:
Yes, looking at it now I'm not even sure why I did it that way 🤔